src/olympia/amo/middleware.py

```diff
@@ -306,3 +306,38 @@ def process_response(self, request, response):
             response['X-AMO-Request-ID'] = request.request_id
 
         return response
+
+
+class CacheControlMiddleware:
+    """Middleware to add Cache-Control: max-age=xxx header to responses that
+    should be cached, Cache-Control: s-maxage:0 to responses that should not.
+
+    The only responses that should be cached are API, unauthenticated responses
+    from "safe" HTTP verbs, or responses that already had a max-age set before
+    being processed by this middleware. In that last case, the Cache-Control
+    header already present is left intact.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        response = self.get_response(request)
+        max_age_from_response = get_max_age(response)
+        request_conditions = (
+            request.is_api
+            and request.method in ('GET', 'HEAD')
+            and 'HTTP_AUTHORIZATION' not in request.META
+            and 'disable_caching' not in request.GET
+        )
+        response_conditions = (
+            not response.cookies
+            and response.status_code >= 200
```

Comment:
This is not new logic (I just moved it) - As you said, since we're only applying that to safe verbs that should be ok. The condition on the status code is really more to ensure everything went as expected.